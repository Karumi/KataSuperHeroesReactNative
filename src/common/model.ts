export class SuperHero {

    public readonly name: string;
    public readonly description: string;
    public readonly isAvenger: boolean;
    public readonly picture?: string;

    public constructor(name: string, description: string, isAvenger: boolean, picture?: string) {
        this.name = name;
        this.description = description;
        this.isAvenger = isAvenger;
        this.picture = picture;
    }
}
