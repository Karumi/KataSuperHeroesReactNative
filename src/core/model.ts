export class SuperHero {

    public readonly name: string;
    public readonly picture: string;
    public readonly isAvenger: boolean;
    public readonly description: string;

    public constructor(name: string, picture: string, isAvenger: boolean, description: string) {
        this.name = name;
        this.picture = picture;
        this.isAvenger = isAvenger;
        this.description = description;
    }
}
