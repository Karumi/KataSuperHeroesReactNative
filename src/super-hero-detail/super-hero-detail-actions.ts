import { Option } from "fp-ts/lib/Option";
import { Action, ActionCreator, Dispatch } from "redux";
import { ThunkAction } from "redux-thunk";
import { action } from "typesafe-actions";
import { SuperHero } from "../core/model";
import { getSuperHeroById } from "../core/super-heroes-repository";
import { CLEAR_SUPER_HERO, LOADING_SUPER_HERO, SUPER_HERO_FETCHED } from "./super-hero-detail-constants";

export const fetchSuperHeroById: ActionCreator<
    ThunkAction<any, any, any, any>> = (superHeroId: string) => (dispatch: Dispatch) => {
        dispatch(loadingSuperHero);
        setTimeout(() => {
            getSuperHeroById(superHeroId).then((sh) => dispatch(superHeroFetched(sh)));
        }, 1000);
    };

export const clearSuperHero: ActionCreator<Action> = () => action(CLEAR_SUPER_HERO);

export const loadingSuperHero: Action =
    action(LOADING_SUPER_HERO);

export const superHeroFetched: ActionCreator<Action> = (hero: Option<SuperHero>) =>
    action(
        SUPER_HERO_FETCHED,
        { superHero: hero },
    );
