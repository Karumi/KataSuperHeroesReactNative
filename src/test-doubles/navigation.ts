
export const mockPlainNavigation = {
    navigate: jest.fn(),
};

export const navigationWithMockedParams = {
    navigate: jest.fn(),
    getParam: jest.fn(),
};
