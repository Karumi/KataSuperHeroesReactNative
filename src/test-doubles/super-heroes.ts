import { SuperHero } from "../core/model";

export const anySuperHeroId: string = "Gambito";

export const anyListOfSuperHeroes: SuperHero[] = [
    new SuperHero("Scarlet Witch",
        "https://i.annihil.us/u/prod/marvel/i/mg/9/b0/537bc2375dfb9.jpg",
        false,
        "Scarlet Witch was born at the Wundagore base of the High Evolutionary, she and her twin "
        + "brother Pietro were the children of Romani couple Django and Marya Maximoff. The "
        + "High Evolutionary supposedly abducted the twins when they were babies and "
        + "experimented on them, once he was disgusted with the results, he returned them to"
        + " Wundagore, disguised as regular mutants."),
    new SuperHero("Iron Man",
        "https://i.annihil.us/u/prod/marvel/i/mg/c/60/55b6a28ef24fa.jpg",
        true,
        "Wounded, captured and forced to build a weapon by his enemies, billionaire "
        + "industrialist Tony Stark instead created an advanced suit of armor to save his "
        + "life and escape captivity. Now with a new outlook on life, Tony uses his money "
        + "and intelligence to make the world a safer, better place as Iron Man."),
    new SuperHero("Wolverine",
        "https://i.annihil.us/u/prod/marvel/i/mg/9/00/537bcb1133fd7.jpg",
        false,
        "Born with super-human senses and the power to heal from almost any wound, " +
        "Wolverine "
        + "was captured by a secret Canadian organization and given an unbreakable "
        + "skeleton and claws. Treated like an animal, it took years for him to control"
        + " himself. Now, he's a premiere member of both the X-Men and the Avengers."),
];

export const anyRegularSuperHero = anyListOfSuperHeroes[0];

export const anyAvenger = anyListOfSuperHeroes[1];
