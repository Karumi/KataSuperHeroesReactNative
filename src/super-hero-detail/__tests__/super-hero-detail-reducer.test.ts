import { Store } from "redux";
import appStore from "../../store";
import { anyRegularSuperHero, anySuperHeroId } from "../../test-doubles/super-heroes";
import { fetchSuperHeroById } from "../super-hero-detail-actions";

jest.mock("../../core/super-heroes-repository", () => {
    return {
        getSuperHeroById: jest.fn(),
    };
});

import { Option, Some } from "fp-ts/lib/Option";
import { SuperHero } from "../../core/model";
import { getSuperHeroById } from "../../core/super-heroes-repository";

it("creates the store using the initial state", () => {
    const store = appStore();

    snapStoreState(store);
});

it("mark the loading property as true while fetching the super hero by id", () => {
    const store = appStore();
    givenTheRepositoryTakesTimeToAnswer();

    store.dispatch(fetchSuperHeroById(anySuperHeroId));

    snapStoreState(store);
});

it("mark the loading property as false and store the super hero fetched", () => {
    const store = appStore();
    givenThereIsASuperHero();

    store.dispatch(fetchSuperHeroById(anySuperHeroId));

    snapStoreState(store);
});

function snapStoreState(store: Store) {
    expect(store.getState().superHeroDetail).toMatchSnapshot();
}

function givenTheRepositoryTakesTimeToAnswer() {

}

function givenThereIsASuperHero() {
    const mock = getSuperHeroById as jest.Mock<Promise<Option<SuperHero>>>;
    
    mock.mockReturnValueOnce(Promise.resolve(new Some(anyRegularSuperHero)));
    console.log("-------> " + JSON.stringify(mock("id")));
}
