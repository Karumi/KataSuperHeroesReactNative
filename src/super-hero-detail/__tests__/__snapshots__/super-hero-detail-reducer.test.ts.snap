// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates the store using the initial state 1`] = `
Object {
  "loading": true,
  "superHero": None {
    "_tag": "None",
  },
}
`;

exports[`mark the loading property as false and store the super hero fetched 1`] = `
Object {
  "loading": true,
  "superHero": None {
    "_tag": "None",
  },
}
`;

exports[`mark the loading property as true while fetching the super hero by id 1`] = `
Object {
  "loading": true,
  "superHero": None {
    "_tag": "None",
  },
}
`;
